nemo_curator/modules/fuzzy_dedup.py

```diff
@@ -475,62 +490,200 @@ def __call__(self, dataset: DocumentDataset):
+        str_id_name: str = "id",
+        bucket_field: str = "_bucket_id",
+        logger: Union[logging.LoggerAdapter, str] = "./",
+    ):
+        """
+        Parameters
+        ----------
+        cache_dir: str or None
+          If specified, will compute & write the edgelist to a file
+        id_fields: list or str
+          id fields of documents in buckets_df
+        str_id_name: str
+          Ignored if there is a single id field. Multiple id fields
+          will be combined into a single id field with the given name.
+        bucket_field: str
+          Column denoting bucket ID
+        num_buckets: Number of bands/buckets to create from the minhash signature.
+          Hashes_per_signature = num_hashes / num_buckets
+        """
+        self.cache_dir = cache_dir
+        self.id_fields = [id_fields] if isinstance(id_fields, str) else id_fields
+        self.str_id_name = str_id_name if len(self.id_fields) > 1 else self.id_fields[0]
+        self.output_ids = [f"{self.str_id_name}_x", f"{self.str_id_name}_y"]
+        self.bucket_field = bucket_field
+        if isinstance(logger, str):
+            self._logger = create_logger(
+                rank=0,
+                log_file=os.path.join(logger, "Buckets_to_Edges.log"),
+                name="Buckets_to_Edges",
+            )
+        else:
+            self._logger = logger
+
+    @staticmethod
+    def _combine_multiple_ids(
+        input_df: dask_cudf.DataFrame, input_id_fields: list, output_id_field: str
+    ) -> dask_cudf.DataFrame:
+        if output_id_field in input_df.columns:
+            raise ValueError(
+                f"Input df already contains column named: {output_id_field}"
+            )
+
+        output_df = input_df.copy()[input_df.columns.difference(input_id_fields)]
+
+        output_df[output_id_field] = input_df[input_id_fields[0]].astype(str)
+        for input_field in input_id_fields[1:]:
+            output_df[output_id_field] = output_df[output_id_field] = (
+                input_df[input_id_fields[0]].astype(str)
+                + "-"
+                + input_df[input_field].astype(str)
+            )
+
+        return output_df
+
+    def buckets_to_edges(
+        self,
+        buckets_df: cudf.DataFrame,
+    ) -> dask_cudf.DataFrame:
+
+        grouped_buckets = (
+            buckets_df.groupby(self.bucket_field)[self.str_id_name]
+            .agg(list)
+            .list.sort_values()
+        )
+        bucket_docs = grouped_buckets.to_arrow().to_pylist()
+        edges = []
+        # Create pairs of all documents within a bucket since they are near duplicates
+        # Effectively create a edge list of all near duplicate documents
+        for bucket_doc in bucket_docs:
+            edges.extend(pairwise(bucket_doc))
+        edges = pd.DataFrame(edges, columns=self.output_ids)
+        edges = pa.Table.from_pandas(edges)
+        result_df = cudf.DataFrame.from_arrow(edges)
```

Comment:
Why do you go through pyarrow instead of using `cudf.from_pandas`?

Reply:
In the grand scheme of things this accounts for a small portion of the method runtime so it's probably insignificant, but for some reason `cudf.from_pandas` is a bit slower than `pandas->arrow->cuDF`. 
Here's some results I got:
```
Time taken from_pandas: 3.225353717803955s
Time taken from_pandas: 3.197599172592163s
Time taken from_pandas: 3.3601927757263184s
Time taken from_pandas: 3.8164901733398438s
Time taken from_pandas: 3.725322961807251s

Time taken from_arrow: 2.4147517681121826s
Time taken from_arrow: 1.6933174133300781s
Time taken from_arrow: 1.6882858276367188s
Time taken from_arrow: 1.3348863124847412s
Time taken from_arrow: 1.6437954902648926s
```

Here's the script:
```python
df_size = 10_000_000
df = pd.DataFrame({"a": np.random.randint(0, np.iinfo("int64").max, df_size), "b": np.random.randint(0, np.iinfo("int64").max, df_size)})
df["a"] = "dataset-1-" + df["a"].astype(str)
df["b"] = "dataset-1-" + df["b"].astype(str)

def f1(df):
    return cudf.from_pandas(df)

def f2(df):
    arrow_df = pa.Table.from_pandas(df)
    return cudf.DataFrame.from_arrow(arrow_df)

n_iter = 5

for i in range(n_iter):
    st = time.time()
    res = f1(df)
    print(f"Time taken from_pandas: {time.time()-st}s")

for i in range(n_iter):
    st = time.time()
    res = f2(df)
    print(f"Time taken from_arrow: {time.time()-st}s")
```

Reply:
Strange, thanks for the details though.